src/librustc_trans/collector.rs

```diff
@@ -304,6 +304,7 @@ fn collect_roots<'a, 'tcx>(scx: &SharedCrateContext<'a, 'tcx>,
         scx.tcx().hir.krate().visit_all_item_likes(&mut visitor);
     }
 
+    roots.retain(|root| root.is_instantiable(scx.tcx()));
```

Comment:
Everywhere except for roots and vtables, instantiability is guaranteed by the program type-checking, so we *want* to produce a clean ICE if the item is not instantiable.